File: frontend/middleware.ts
import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';

export function middleware(request: NextRequest) {
  const nonce = Buffer.from(crypto.randomUUID()).toString('base64');

  const scriptSrc = ["'self'", `'nonce-${nonce}'`, "'strict-dynamic'", 'https:'];
  const connectSrc = new Set(["'self'"]);

  const configuredApiUrl = process.env.NEXT_PUBLIC_API_URL ?? 'https://api.valsa.solutions';
  const candidateUrls = [
    configuredApiUrl,
    process.env.NEXT_PUBLIC_WS_URL,
    process.env.NEXT_PUBLIC_INTERNAL_API_URL,
    'https://edustream.valsa.solutions',
    'https://api.valsa.solutions',
    'https://minio.valsa.solutions',
  ].filter((value): value is string => Boolean(value));

  for (const url of candidateUrls) {
    try {
      const origin = new URL(url).origin;
      connectSrc.add(origin);
      
      // Add WebSocket equivalent for HTTPS origins
      if (origin.startsWith('https://')) {
        const wsOrigin = origin.replace('https://', 'wss://');
        connectSrc.add(wsOrigin);
      }
    } catch {
      // Ignore invalid URLs
    }
  }

  const cspHeader = `
    default-src 'self';
    script-src ${scriptSrc.join(' ')};
    connect-src ${Array.from(connectSrc).join(' ')};
    style-src 'self' 'unsafe-inline' https://fonts.googleapis.com;
    img-src 'self' blob: data: https:;
    font-src 'self' https://fonts.gstatic.com;
    object-src 'none';
    base-uri 'self';
    form-action 'self';
    frame-ancestors 'none';
  `.replace(/\s{2,}/g, ' ').trim();

  const requestHeaders = new Headers(request.headers);
  requestHeaders.set('x-nonce', nonce);
  requestHeaders.set('Content-Security-Policy', cspHeader);

  const response = NextResponse.next({
    request: {
      headers: requestHeaders,
    },
  });

  response.headers.set('Content-Security-Policy', cspHeader);
  response.headers.set('X-Content-Type-Options', 'nosniff');
  response.headers.set('X-Frame-Options', 'DENY');
  response.headers.set('X-XSS-Protection', '1; mode=block');
  response.headers.set('Referrer-Policy', 'strict-origin-when-cross-origin');
  response.headers.set('Permissions-Policy', 'camera=(), microphone=(), geolocation=()');

  return response;
}

export const config = {
  matcher: [
    '/((?!api|_next/static|_next/image|favicon.ico|icon.svg|apple-icon.svg).*)',
  ],
};

File: backend/core/middleware.py
"""Security middleware for FastAPI application."""

from typing import Callable

from fastapi import Request, Response
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.types import ASGIApp


class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """
    Middleware to add security headers to all responses.

    Following OWASP best practices and CLAUDE.md security guidelines.
    """

    def __init__(self, app: ASGIApp):
        super().__init__(app)

    async def dispatch(self, request: Request, call_next: Callable) -> Response:
        """Add security headers to response."""
        response = await call_next(request)

        # Relax CSP for OpenAPI docs to avoid breaking Swagger UI assets
        if request.url.path.startswith(("/docs", "/redoc", "/openapi", "/static")):
            response.headers["Content-Security-Policy"] = (
                "default-src 'self'; "
                "img-src 'self' data: https:; "
                "style-src 'self' 'unsafe-inline'; "
                "script-src 'self' https: 'unsafe-inline'; "
                "connect-src 'self'; "
                "frame-ancestors 'none'; "
                "base-uri 'self'; "
                "form-action 'self';"
            )
        else:
            response.headers["Content-Security-Policy"] = (
                "default-src 'none'; "
                "connect-src 'self'; "
                "img-src 'self' data:; "
                "script-src 'none'; "
                "style-src 'none'; "
                "font-src 'none'; "
                "object-src 'none'; "
                "frame-ancestors 'none'; "
                "base-uri 'self'; "
                "form-action 'self';"
            )

        # Prevent clickjacking attacks
        response.headers["X-Frame-Options"] = "DENY"

        # Prevent MIME type sniffing
        response.headers["X-Content-Type-Options"] = "nosniff"

        # Enable browser XSS protection
        response.headers["X-XSS-Protection"] = "1; mode=block"

        # Referrer Policy - Control referrer information
        response.headers["Referrer-Policy"] = "strict-origin-when-cross-origin"

        # Permissions Policy (formerly Feature Policy)
        response.headers["Permissions-Policy"] = (
            "accelerometer=(), "
            "camera=(), "
            "geolocation=(), "
            "gyroscope=(), "
            "magnetometer=(), "
            "microphone=(), "
            "payment=(), "
            "usb=()"
        )

        # Strict Transport Security - Enforce HTTPS (only in production)
        # Uncomment when deploying to production with HTTPS
        # response.headers["Strict-Transport-Security"] = (
        #     "max-age=31536000; includeSubDomains; preload"
        # )

        # Remove server header to avoid information disclosure
        if "Server" in response.headers:
            del response.headers["Server"]

        return response
